resolver/resolver.cc

```diff
@@ -1598,80 +1626,84 @@ class ResolveConstantsWalk {
         bool progress = true;
         bool first = true; // we need to run at least once to force class aliases and type aliases
 
-        while (progress && (first || !todo.empty() || !todoAncestors.empty())) {
-            first = false;
-            counterInc("resolve.constants.retries");
-            {
-                Timer timeit(gs.tracer(), "resolver.resolve_constants.fixed_point.ancestors");
-                // This is an optimization. The order should not matter semantically
-                // We try to resolve most ancestors second because this makes us much more likely to resolve
-                // everything else.
-                long retries = 0;
-                auto f = [&gs, &retries](ResolveItems<AncestorResolutionItem> &job) -> bool {
-                    core::MutableContext ctx(gs, core::Symbols::root(), job.file);
-                    const auto origSize = job.items.size();
-                    auto g = [&](AncestorResolutionItem &item) -> bool {
-                        auto resolved = resolveAncestorJob(ctx, item, false);
-                        if (resolved) {
-                            tryRegisterSealedSubclass(ctx, item);
-                        }
-                        return resolved;
+        // If the constant didn't immediately resolve during the initial treewalk, and we're not
+        // allowed to mutate GlobalState, it will never resolve. Let's just skip to the error phase.
+        if constexpr (isMutableStateType) {
```

Comment:
This is the key insight that makes this change workable at all, and not immeasurably hairy.

The ResolveConstantsWalk will try to avoid enqueuing work into the various todo lists for the sake of speed. That means that if there's anything in the todo lists at the end of the loop, they're there because they were hoping to be resolved / processed once some other todo list item had been processed.

But we won't actually be able to make progress. Making progress requires being able to modify the symbol table, so that other items can discover new state in the symbol table and make their own progress. That will never happen in the immutable resolver mode, so we can skip straight to the final part.